### PR TITLE
RUM-15277: Propagate native anonymousId to webview events

### PR DESCRIPTION
**Pull request by kikoveiga**
### What does this PR do?

Propagates the native `anonymousId` into webview RUM and Log events. Even if browser set it on their side, we will overwrite it (we do the same with `applicationId`, for instance).

iOS counterpart: https://github.com/DataDog/dd-sdk-ios/pull/2847

### Motivation

[RUMS-5561](https://datadoghq.atlassian.net/browse/RUMS-5661)

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](../CONTRIBUTING.md) doc)

[RUMS-5561]: https://datadoghq.atlassian.net/browse/RUMS-5561?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by 0xnm — DISMISSED** [review]

File: features/dd-sdk-android-webview/src/main/kotlin/com/datadog/android/webview/internal/rum/WebViewRumEventMapper.kt [hunk 1]
```diff
@@ -64,6 +66,8 @@ internal class WebViewRumEventMapper(
             event.add(SESSION_KEY_NAME, session)
         }
 
+        event.add(USR_KEY_NAME, userInfo.toJson())
```
[the remark this answers is not in the document]
Comment (kikoveiga) [reply features/dd-sdk-android-webview/src/main/kotlin/com/datadog/android/webview/internal/rum/WebViewRumEventMapper.kt]: Fixed.

File: features/dd-sdk-android-webview/src/main/kotlin/com/datadog/android/webview/internal/log/WebViewLogEventConsumer.kt [hunk 2]
```diff
@@ -53,13 +53,18 @@ internal class WebViewLogEventConsumer(
...
+        addUserInfo(event, datadogContext)
         if (rumContext != null) {
             event.addProperty(LogAttributes.RUM_APPLICATION_ID, rumContext.applicationId)
             event.addProperty(LogAttributes.RUM_SESSION_ID, rumContext.sessionId)
         }
         return event
     }
 
+    private fun addUserInfo(event: JsonObject, datadogContext: DatadogContext) {
+        event.add(USR_KEY_NAME, datadogContext.userInfo.toJson())
```
[the remark this answers is not in the document]
Comment (kikoveiga) [reply features/dd-sdk-android-webview/src/main/kotlin/com/datadog/android/webview/internal/log/WebViewLogEventConsumer.kt]: Fixed.

**Review by 0xnm — DISMISSED** [review]

File: features/dd-sdk-android-webview/src/main/kotlin/com/datadog/android/webview/internal/log/WebViewLogEventConsumer.kt [hunk 3]
```diff
@@ -53,13 +53,24 @@ internal class WebViewLogEventConsumer(
...
             event.addProperty(LogAttributes.RUM_APPLICATION_ID, rumContext.applicationId)
             event.addProperty(LogAttributes.RUM_SESSION_ID, rumContext.sessionId)
         }
         return event
     }
 
+    private fun addUserInfo(event: JsonObject, datadogContext: DatadogContext) {
+        val usr = event.getAsJsonObject(USR_KEY_NAME) ?: JsonObject()
+        for ((key, value) in datadogContext.userInfo.toJson().asJsonObject.entrySet()) {
+            if (!usr.has(key)) {
```
Comment (ambushwork) [inline features/dd-sdk-android-webview/src/main/kotlin/com/datadog/android/webview/internal/log/WebViewLogEventConsumer.kt]: Question: The logic seems like it overrides only a part of the user info if they are absent and keep the rest of them if present, what's the reason of doing that?

**Review by ambushwork — DISMISSED** [review]

File: features/dd-sdk-android-webview/src/main/kotlin/com/datadog/android/webview/internal/log/WebViewLogEventConsumer.kt (hunk 3 above)
Comment (kikoveiga) [reply features/dd-sdk-android-webview/src/main/kotlin/com/datadog/android/webview/internal/log/WebViewLogEventConsumer.kt]: The logic here is that we do not override the information that is already present from the browser side, and just add new fields that are not there (`anonymousId` will be the case for example). I added an image on the pr description for some clarity.

Comment (ambushwork) [reply features/dd-sdk-android-webview/src/main/kotlin/com/datadog/android/webview/internal/log/WebViewLogEventConsumer.kt]: This sounds strange for me, in some corner cases when Browser SDK and native SDK both possess partial user info can lead to a fake composed user info if I am not wrong.

```
// Webview
UserInfo(
    anonymousId = null,
    id = "001" ,
    name = "John" ,
    email = null
) 

// Native
UserInfo(
    anonymousId = "anonymous name",
    id = "002" ,
    name = "Bob" ,
    email = "Bob@datadoghq.com"
) 

// After the logic it becomes
UserInfo(
    anonymousId = "anonymous name",
    id = "001" ,
    name = "John" ,
    email = "Bob@datadoghq.com"
) 

```

Comment (kikoveiga) [reply features/dd-sdk-android-webview/src/main/kotlin/com/datadog/android/webview/internal/log/WebViewLogEventConsumer.kt]: Oh ok I didn't understand previously, thanks! So maybe I should just stick to adding the anonymousId, which is a native field only, wdyt?

Comment (ambushwork) [reply features/dd-sdk-android-webview/src/main/kotlin/com/datadog/android/webview/internal/log/WebViewLogEventConsumer.kt]: it can be an option, or the alternative solution is to completely override the user info of Browser SDK. it's better to check with iOS what is their approach.

Comment (kikoveiga) [reply features/dd-sdk-android-webview/src/main/kotlin/com/datadog/android/webview/internal/log/WebViewLogEventConsumer.kt]: iOS doesn't do it yet as well, I'm supposed to do the same there: [RUM-15279](https://datadoghq.atlassian.net/browse/RUM-15279).

I had tried before to override everything, but it didn't work because some fields disappeared (web-side name and email for example). Tested it in Shopist. Also, codex advised against this in the above resolved comments.

Comment (ambushwork) [reply features/dd-sdk-android-webview/src/main/kotlin/com/datadog/android/webview/internal/log/WebViewLogEventConsumer.kt]: I think codex missed the whole picture when giving the advise, IMO the correct path should be:
* If we detect that two user info are same, by checking id for example, we are safe to merge them
* If not, we need to make a decision that if we want to keep the native one or browser one.

Also it's better to add a unit test to protect this corner case.

Comment (kikoveiga) [reply features/dd-sdk-android-webview/src/main/kotlin/com/datadog/android/webview/internal/log/WebViewLogEventConsumer.kt]: I don't know the exact logic of webviews and browser sdk, but isn't the id set on their side when on a webview? If so, ids will never be the same. I think info from browser and from native is unrelated, and trying to merge them can change behaviour unexpectedly. That's why I was only adding non-existing fields, but the safest is definitely just adding the `anonymousId` (which is only present on the native side).

Comment (cdn34dd) [reply features/dd-sdk-android-webview/src/main/kotlin/com/datadog/android/webview/internal/log/WebViewLogEventConsumer.kt]: The `anonymousId` field is not a mobile field only, you can also have it the browser SDK, I think the difference is that in the browser SDK is disabled by default and we enable it by default on mobile if I'm not mistaken.

I'd say here we should simply add the `anonymousId` if not present to ensure we don't make any breaking changes in minor versions as people can be depending on things working like they do now. If we want to review this at some point we should do so in a major version.

**Review by cdn34dd — COMMENTED** [review]
I notice that on webview events context also has a `usr` field, and it seems that we're not setting `anonymousId` there:
```
{
  (...)
  "context": {
    "usr": {
      "use_new_checkout_flow": true,
      "name": "harper.harry",
      "id": "1gfdsgsdfg",
      "team": "engineering",
      "use_discount": false,
      "email": "harper.harry@example.com"
    },
    "browser_test": false
  }
 (...)
}
```
I'm not sure if there's any backend functionality that specifically looks for `context.usr`, but maybe the safe option would be to add it also there ? As it seems that if browser were to set `anonymousId` itself, this would also be there.

**Review by 0xnm — DISMISSED** [review]

**Review by 0xnm — APPROVED** [review]
